File: node/src/runtime/durable-content.ts
/** Provider-neutral durable content ABI.
 *
 * This is deliberately separate from provider-facing ContentPart. Durable records carry only
 * portable content and explicit payload locators; protocol adapters decide how to serialize it.
 */

export class DurableContentError extends Error {
  constructor(message: string) {
    super(message)
    this.name = "DurableContentError"
  }
}

export type DurableSource =
  | { kind: "url"; url: string }
  | { kind: "base64"; data: string }
  | { kind: "file_id"; id: string; affinity: { provider_id: string; endpoint_id: string } }
  | { kind: "object"; handle: string; owner: string; payload_ref: string }

export type DurableContentBlock =
  | { type: "text"; text: string }
  | { type: "image" | "audio" | "video" | "file"; source: DurableSource; media_type?: string; provider_options?: Record<string, unknown> }

export interface DurableContent {
  blocks: DurableContentBlock[]
}

export interface DurableToolResult extends DurableContent {
  call_id: string
  is_error: boolean
}

import type { ToolOutputBlock } from "../types.js"

const CONTENT_KEYS = new Set(["blocks"])
const TOOL_RESULT_KEYS = new Set(["call_id", "is_error", "blocks"])
const SOURCE_KEYS: Record<DurableSource["kind"], Set<string>> = {
  url: new Set(["kind", "url"]),
  base64: new Set(["kind", "data"]),
  file_id: new Set(["kind", "id", "affinity"]),
  object: new Set(["kind", "handle", "owner", "payload_ref"]),
}

function object(value: unknown, label: string): Record<string, unknown> {
  if (!value || typeof value !== "object" || Array.isArray(value)) throw new DurableContentError(`${label} must be an object`)
  return value as Record<string, unknown>
}

function exactKeys(value: Record<string, unknown>, allowed: Set<string>, label: string): void {
  for (const key of Object.keys(value)) if (!allowed.has(key)) throw new DurableContentError(`${label} has unknown field ${key}`)
}

function requiredString(value: unknown, label: string): string {
  if (typeof value !== "string" || value.length === 0) throw new DurableContentError(`${label} must be a non-empty string`)
  return value
}

function requiredBoolean(value: unknown, label: string): boolean {
  if (typeof value !== "boolean") throw new DurableContentError(`${label} must be a boolean`)
  return value
}

function base64(value: unknown, label: string): string {
  const data = requiredString(value, label)
  if (data.length % 4 !== 0 || !/^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$/.test(data)) {
    throw new DurableContentError(`${label} is not valid base64`)
  }
  return data
}

function source(value: unknown, label: string): DurableSource {
  const raw = object(value, `${label} source`)
  const kind = raw.kind
  if (kind !== "url" && kind !== "base64" && kind !== "file_id" && kind !== "object") throw new DurableContentError(`${label} source kind is invalid`)
  exactKeys(raw, SOURCE_KEYS[kind], `${label} source`)
  if (kind === "url") return { kind, url: requiredString(raw.url, `${label} URL`) }
  if (kind === "base64") return { kind, data: base64(raw.data, `${label} base64 data`) }
  if (kind === "file_id") {
    const id = requiredString(raw.id, `${label} file id`)
    const a = object(raw.affinity, `${label} affinity`)
    exactKeys(a, new Set(["provider_id", "endpoint_id"]), `${label} affinity`)
    return {
      kind,
      id,
      affinity: {
        provider_id: requiredString(a.provider_id, `${label} affinity provider_id`),
        endpoint_id: requiredString(a.endpoint_id, `${label} affinity endpoint_id`),
      },
    }
  }
  return {
    kind,
    handle: requiredString(raw.handle, `${label} object handle`),
    owner: requiredString(raw.owner, `${label} object owner`),
    payload_ref: requiredString(raw.payload_ref, `${label} payload_ref`),
  }
}

function block(value: unknown, index: number): DurableContentBlock {
  const raw = object(value, `content block ${index}`)
  const type = raw.type
  if (type === "text") {
    exactKeys(raw, new Set(["type", "text"]), `content block ${index}`)
    return { type, text: typeof raw.text === "string" ? raw.text : (() => { throw new DurableContentError(`content block ${index} text must be a string`) })() }
  }
  if (type === "tool_result") throw new DurableContentError("nested tool_result blocks are forbidden")
  if (type !== "image" && type !== "audio" && type !== "video" && type !== "file") throw new DurableContentError(`unknown content block type: ${String(type)}`)
  exactKeys(raw, new Set(["type", "source", "media_type", "provider_options"]), `content block ${index}`)
  const mediaType = raw.media_type === undefined ? undefined : requiredString(raw.media_type, `content block ${index} media_type`)
  let providerOptions: Record<string, unknown> | undefined
  if (raw.provider_options !== undefined) providerOptions = object(raw.provider_options, `content block ${index} provider_options`)
  return {
    type,
    source: source(raw.source, `content block ${index}`),
    ...(mediaType === undefined ? {} : { media_type: mediaType }),
    ...(providerOptions === undefined ? {} : { provider_options: providerOptions }),
  }
}

function blocks(value: unknown): DurableContentBlock[] {
  if (!Array.isArray(value)) throw new DurableContentError("content blocks must be an array")
  return value.map(block)
}

export function decodeDurableContent(value: unknown): DurableContent {
  const raw = object(value, "durable content")
  exactKeys(raw, CONTENT_KEYS, "durable content")
  return { blocks: blocks(raw.blocks) }
}

export function decodeDurableToolResult(value: unknown): DurableToolResult {
  const raw = object(value, "durable tool result")
  exactKeys(raw, TOOL_RESULT_KEYS, "durable tool result")
  return { call_id: requiredString(raw.call_id, "tool result call_id"), is_error: requiredBoolean(raw.is_error, "tool result is_error"), blocks: blocks(raw.blocks) }
}

export function encodeDurableContent(content: DurableContent): Record<string, unknown> {
  const decoded = decodeDurableContent(content)
  return { blocks: decoded.blocks }
}

export function encodeDurableToolResult(result: DurableToolResult): Record<string, unknown> {
  const decoded = decodeDurableToolResult(result)
  return { call_id: decoded.call_id, is_error: decoded.is_error, blocks: decoded.blocks }
}

export function toolOutputBlocksToDurable(blocks: readonly ToolOutputBlock[]): DurableContentBlock[] {
  return blocks.map((part, index) => {
    if (part.type === "text") return { type: "text", text: part.text }
    const source = part.source
    const durableSource = source.kind === "fileId"
      ? {
          kind: "file_id" as const,
          id: source.id,
          affinity: source.affinity
            ? { provider_id: source.affinity.providerId, endpoint_id: source.affinity.endpointId }
            : (() => { throw new DurableContentError(`tool output block ${index} fileId source requires endpoint affinity`) })(),
        }
      : source.kind === "base64" ? { kind: "base64" as const, data: source.data }
      : source.kind === "url" ? { kind: "url" as const, url: source.url }
      : {
          kind: "object" as const,
          handle: source.handle,
          owner: source.owner ?? (() => { throw new DurableContentError(`tool output block ${index} object source requires owner`) })(),
          payload_ref: source.payloadRef ?? (() => { throw new DurableContentError(`tool output block ${index} object source requires payloadRef`) })(),
        }
    return { type: part.type, source: durableSource, ...(part.mediaType ? { media_type: part.mediaType } : {}), ...(part.providerOptions ? { provider_options: part.providerOptions } : {}) }
  })
}

export function durableBlocksToToolOutput(blocks: readonly DurableContentBlock[]): ToolOutputBlock[] {
  return blocks.map((part) => {
    if (part.type === "text") return part
    const source = part.source
    const sdkSource = source.kind === "file_id"
      ? { kind: "fileId" as const, id: source.id, affinity: { providerId: source.affinity.provider_id, endpointId: source.affinity.endpoint_id } }
      : source.kind === "base64" ? source
      : source.kind === "url" ? source
      : { kind: "object" as const, handle: source.handle, owner: source.owner, payloadRef: source.payload_ref }
    return { type: part.type, source: sdkSource, ...(part.media_type ? { mediaType: part.media_type } : {}), ...(part.provider_options ? { providerOptions: part.provider_options } : {}) } as ToolOutputBlock
  })
}

File: crates/deepstrike-core/src/types/durable_content.rs
//! Provider-neutral content used by durable events/checkpoints.
//! Protocol adapters own vendor serialization; this DTO owns only portable content and payload
//! locator semantics. Unknown fields and nested tool results fail closed.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct DurableContent {
    pub blocks: Vec<DurableContentBlock>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct DurableToolResult {
    pub call_id: String,
    pub is_error: bool,
    pub blocks: Vec<DurableContentBlock>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case", deny_unknown_fields)]
pub enum DurableContentBlock {
    Text {
        text: String,
    },
    Image {
        source: DurableSource,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        media_type: Option<String>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        provider_options: Option<serde_json::Value>,
    },
    Audio {
        source: DurableSource,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        media_type: Option<String>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        provider_options: Option<serde_json::Value>,
    },
    Video {
        source: DurableSource,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        media_type: Option<String>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        provider_options: Option<serde_json::Value>,
    },
    File {
        source: DurableSource,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        media_type: Option<String>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        provider_options: Option<serde_json::Value>,
    },
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "snake_case", deny_unknown_fields)]
pub enum DurableSource {
    Url {
        url: String,
    },
    Base64 {
        data: String,
    },
    FileId {
        id: String,
        affinity: EndpointAffinity,
    },
    Object {
        handle: String,
        owner: String,
        payload_ref: String,
    },
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct EndpointAffinity {
    pub provider_id: String,
    pub endpoint_id: String,
}

impl DurableContent {
    pub fn text(text: impl Into<String>) -> Self {
        Self {
            blocks: vec![DurableContentBlock::Text { text: text.into() }],
        }
    }

    pub fn decode(value: serde_json::Value) -> Result<Self, serde_json::Error> {
        let content: Self = serde_json::from_value(value)?;
        content.validate()?;
        Ok(content)
    }

    pub fn validate(&self) -> Result<(), serde_json::Error> {
        for block in &self.blocks {
            validate_block(block)?;
        }
        Ok(())
    }
}

impl DurableToolResult {
    pub fn text(call_id: impl Into<String>, output: impl Into<String>, is_error: bool) -> Self {
        Self {
            call_id: call_id.into(),
            is_error,
            blocks: vec![DurableContentBlock::Text {
                text: output.into(),
            }],
        }
    }

    pub fn decode(value: serde_json::Value) -> Result<Self, serde_json::Error> {
        let result: Self = serde_json::from_value(value)?;
        result.validate()?;
        Ok(result)
    }

    pub fn validate(&self) -> Result<(), serde_json::Error> {
        require_non_empty(&self.call_id, "tool result call_id")?;
        for block in &self.blocks {
            validate_block(block)?;
        }
        Ok(())
    }
}

fn invalid(message: impl Into<String>) -> serde_json::Error {
    serde_json::Error::io(std::io::Error::new(
        std::io::ErrorKind::InvalidData,
        message.into(),
    ))
}

fn require_non_empty(value: &str, label: &str) -> Result<(), serde_json::Error> {
    if value.is_empty() {
        return Err(invalid(format!("{label} must be a non-empty string")));
    }
    Ok(())
}

fn validate_block(block: &DurableContentBlock) -> Result<(), serde_json::Error> {
    match block {
        DurableContentBlock::Text { .. } => Ok(()),
        DurableContentBlock::Image {
            source, media_type, ..
        }
        | DurableContentBlock::Audio {
            source, media_type, ..
        }
        | DurableContentBlock::Video {
            source, media_type, ..
        }
        | DurableContentBlock::File {
            source, media_type, ..
        } => {
            validate_source(source)?;
            if let Some(media_type) = media_type {
                require_non_empty(media_type, "content media_type")?;
            }
            Ok(())
        }
    }
}

fn validate_source(source: &DurableSource) -> Result<(), serde_json::Error> {
    match source {
        DurableSource::Url { url } => require_non_empty(url, "content URL"),
        DurableSource::Base64 { data } => {
            require_non_empty(data, "content base64 data")?;
            if !valid_standard_base64(data) {
                return Err(invalid("content base64 data is not valid base64"));
            }
            Ok(())
        }
        DurableSource::FileId { id, affinity } => {
            require_non_empty(id, "content file id")?;
            require_non_empty(&affinity.provider_id, "content file affinity provider_id")?;
            require_non_empty(&affinity.endpoint_id, "content file affinity endpoint_id")
        }
        DurableSource::Object {
            handle,
            owner,
            payload_ref,
        } => {
            require_non_empty(handle, "content object handle")?;
            require_non_empty(owner, "content object owner")?;
            require_non_empty(payload_ref, "content object payload_ref")
        }
    }
}

fn valid_standard_base64(data: &str) -> bool {
    let bytes = data.as_bytes();
    if bytes.len() % 4 != 0 {
        return false;
    }
    let padding = bytes.iter().rev().take_while(|&&byte| byte == b'=').count();
    if padding > 2 || padding > bytes.len() {
        return false;
    }
    bytes[..bytes.len().saturating_sub(padding)]
        .iter()
        .all(|byte| byte.is_ascii_alphanumeric() || *byte == b'+' || *byte == b'/')
        && bytes[bytes.len().saturating_sub(padding)..]
            .iter()
            .all(|byte| *byte == b'=')
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn versioned_and_output_only_tool_results_are_rejected() {
        assert!(
            DurableToolResult::decode(
                json!({"schema_version":1,"call_id":"c1","is_error":false,"blocks":[]})
            )
            .is_err()
        );
        assert!(
            DurableToolResult::decode(json!({"call_id":"c1","output":"ok","is_error":false}))
                .is_err()
        );
    }

    #[test]
    fn unknown_and_nested_blocks_are_rejected() {
        assert!(
            DurableContent::decode(json!({"blocks":[{"type":"text","text":"x","extra":true}]}))
                .is_err()
        );
        assert!(
            DurableContent::decode(json!({"blocks":[{"type":"tool_result","call_id":"nested"}]}))
                .is_err()
        );
        assert!(DurableContent::decode(json!({"blocks":[{"type":"file","source":{"kind":"object","handle":"h","owner":"host"}}]})).is_err());
        assert!(DurableContent::decode(json!({"schema_version":1,"blocks":[]})).is_err());
        assert!(
            DurableContent::decode(
                json!({"blocks":[{"type":"file","source":{"kind":"file_id","id":"f"}}]})
            )
            .is_err()
        );
    }

    #[test]
    fn media_affinity_and_payload_ownership_round_trip() {
        let content = DurableContent {
            blocks: vec![
                DurableContentBlock::File {
                    source: DurableSource::FileId {
                        id: "f1".into(),
                        affinity: EndpointAffinity {
                            provider_id: "openai".into(),
                            endpoint_id: "responses".into(),
                        },
                    },
                    media_type: Some("application/pdf".into()),
                    provider_options: None,
                },
                DurableContentBlock::Video {
                    source: DurableSource::Object {
                        handle: "h1".into(),
                        owner: "host".into(),
                        payload_ref: "sha256:abc".into(),
                    },
                    media_type: Some("video/mp4".into()),
                    provider_options: None,
                },
            ],
        };
        let bytes = serde_json::to_value(&content).unwrap();
        assert_eq!(DurableContent::decode(bytes).unwrap(), content);
    }

    #[test]
    fn shared_canonical_tool_result_fixture_round_trips() {
        let fixture =
            include_str!("../../../../tests/fixtures/durable-content/canonical-tool-result.json");
        let value: serde_json::Value = serde_json::from_str(fixture).unwrap();
        let decoded = DurableToolResult::decode(value.clone()).unwrap();
        assert_eq!(serde_json::to_value(decoded).unwrap(), value);
    }
}
